stable_baselines/common/policies.py

```diff
@@ -105,24 +105,59 @@ class BasePolicy(ABC):
     :param add_action_ph: (bool) whether or not to create an action placeholder
     """
 
+    stateful = False
```

Comment:
I also feel queazy about using a static class variable, but the problem is we sometimes need to check if a *class* is stateful before creating the policy. For example: https://github.com/hill-a/stable-baselines/blob/f564519f56b9a7e72b33be38415d0ea123460884/stable_baselines/common/base_class.py#L98

The only alternative I can think of would be to make abstract classes `StatelessPolicy` and `StatefulPolicy` that subclass `BasePolicy`, and we can then check with `issubclass`. Also open to other suggestions.

Reply:
I see you point. I looked deeper into the issue.

**Use functors instead of classes**: We are using policy classes like a functor. Maybe we should make policies functors? Then `stateful` can be a property. It would also fix another ugly flaw in the current policy implementation: The weight sharing between act and run models works solely based on the assumption that the same policy class is instantiated twice (once for run and once for learning) in the same `tf.variable_scope`. This is *very* brittle! 

**Examples**: What if we for some reason instantiate two different policies whose `tf.Variable` names overlap in the same scope? Suddenly they start sharing weights! 
What if the designer of an RL algorithm decides to put run and learn models in different scopes? Without looking deep into the tf graph she won't figure out why the run policy never improves.
What if there is one day a RL algorithm with dueling policies? We need to be aware that they need to be put in different variable scopes otherwise they start sharing weights and the duel is meaningless.

With a functor design we could ensure that weight sharing happens *iff* two models graphs are constructed by the same functor completely independent of any external variable scope.

What do you guys think? I am in favor of fixing the old underlying issues rather than adding more soso constructions to work around the underlying issues but I also understand that this involves a lot of change (possibly breaking backwards compatibility) and work.

Reply:
Have you a concrete (simple example, maybe a code sketch) of your idea?
(I understand the issues, not sure to understand the proposed solution)

Reply:
Also not sure I follow the details of @erniejunior's solution. In particular:
  - For the property, would this be a function property of the functor? I agree this is conceptually clearer but I'm not sure it'll be easier for user's to reason about than a static class property.
  - The variable sharing seems like a big problem, but how would this work with the functor? Do we have some factory that creates functors? I might want to create two policies that don't share weights (e.g. in multi-agent RL), would I have to create two different functors to do this?

Reply:
OK I try a little mock-up with a stateless deterministic MLP model. The value and policy network share the hidden layer.
```python
import tensorflow as tf
import numpy as np


class BasePolicy:
    def __init__(self, sess):
        self.sess = sess

    @property
    def stateless(self):
        return self.initial_state is None

    @property
    def initial_state(self):
        raise NotImplementedError()

    def make_state_placeholder(self, batch_size):
        assert not self.stateless
        return tf.placeholder(tf.float32, (batch_size,) + self.initial_state.shape)


class MyPolicy(BasePolicy):
    def __init__(self, sess, obs_size, hidden_size, action_size):
        super().__init__(self, sess)
        self.w_common = tf.Variable(np.random.rand(obs_size, hidden_size))
        self.b_common = tf.Variable(np.zeros(hidden_size))

        self.w_pi = tf.Variable(np.random.rand(hidden_size, action_size))
        self.b_pi = tf.Variable(np.zeros(action_size))

        self.w_vf = tf.Variable(np.random.rand(hidden_size, 1))
        self.b_vf = tf.Variable(np.zeros(1))

    @property
    def initial_state(self):
        return None

    def __call__(self, observations_ph, states_ph):
        hidden_act = tf.nn.relu(tf.matmul(observations_ph, self.w_common) + self.b_common)

        action = tf.nn.relu(tf.matmul(hidden_act, self.w_pi) + self.b_pi)
        vf = tf.nn.relu(tf.matmul(hidden_act, self.w_vf) + self.b_vf)

        step_fn = lambda observations, states: self.sess.run(action, dict(observations_ph=observations))
        return action, vf, step_fn
```
Note that this is just one way of many to handle the 'step' mechanics. I could also think of a generic `ModelStepper` that gets as arguments the observation and state placeholders as well as the `action` tensor. This model stepper is generic and works with any policy (right now all the `step` methods do look rather similar anyways, we have a lot of code duplication).

Also in practice we probably need to handle probabilistic policies, observation preprocessing etc. This can all be done in the Base policy.

With this setup we would pass an instance of `MyPolicy` to the RL algorithm instead of the `MyPolicy` class. Feel free to ask any specific questions if you need clarification. I try to answer them quickly this time.